In [84]:
#libraries
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


In [85]:
#libraries
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import CountVectorizer

import seaborn as sns
import matplotlib.pyplot as plt

from tqdm.auto import tqdm
import time

In [86]:
# Load the dataset
data = pd.read_csv('C:\\Users\\Priyanka Sawant\\Desktop\\Python\\Exam NLP\\IMDB Dataset.csv')

In [87]:
data.shape

(50000, 2)

In [88]:
data['review'][3]

"Basically there's a family where a little boy (Jake) thinks there's a zombie in his closet & his parents are fighting all the time.<br /><br />This movie is slower than a soap opera... and suddenly, Jake decides to become Rambo and kill the zombie.<br /><br />OK, first of all when you're going to make a film you must Decide if its a thriller or a drama! As a drama the movie is watchable. Parents are divorcing & arguing like in real life. And then we have Jake with his closet which totally ruins all the film! I expected to see a BOOGEYMAN similar movie, and instead i watched a drama with some meaningless thriller spots.<br /><br />3 out of 10 just for the well playing parents & descent dialogs. As for the shots with Jake: just ignore them."

In [89]:
data['review'][34]

"I watched this film not really expecting much, I got it in a pack of 5 films, all of which were pretty terrible in their own way for under a fiver so what could I expect? and you know what I was right, they were all terrible, this movie has a few (and a few is stretching it) interesting points, the occasional camcorder view is a nice touch, the drummer is very like a drummer, i.e damned annoying and, well thats about it actually, the problem is that its just so boring, in what I can only assume was an attempt to build tension, a whole lot of nothing happens and when it does its utterly tedious (I had my thumb on the fast forward button, ready to press for most of the movie, but gave it a go) and seriously is the lead singer of the band that great looking, coz they don't half mention how beautiful he is a hell of a lot, I thought he looked a bit like a meercat, all this and I haven't even mentioned the killer, I'm not even gonna go into it, its just not worth explaining. Anyway as far 

In [90]:
data.columns

Index(['review', 'sentiment'], dtype='object')

In [91]:
#count of positive &  negative reviews sentiment
data['sentiment'].value_counts()

positive    25000
negative    25000
Name: sentiment, dtype: int64

In [92]:
#check null values
data.isnull().sum()

review       0
sentiment    0
dtype: int64

In [93]:
#duplicate text
data.duplicated().sum()

418

In [94]:
#drop suplicate text
data.drop_duplicates(inplace=True)

In [95]:
data.duplicated().sum()

0

# Text Preprocessing

#### Q.1 Preprocess the text (i.e., the steps required prior to converting the sentence into a vector) using any library of your choice

# Lowercasing
convert NLP -> nlp

In [96]:
data['review'] = data['review'].str.lower()

In [97]:
data.head(3)

review sentiment
0  one of the other reviewers has mentioned that ...  positive
1  a wonderful little production. <br /><br />the...  positive
2  i thought this was a wonderful way to spend ti...  positive

# Remove HTML Tags

Remove all html tags(such as <br,<li, etc) from text

In [98]:
import re
def remove_html_tags(text):
    pattern = re.compile('<.*?>')
    return pattern.sub(f'',text)

In [99]:
data['review']=data['review'].apply(remove_html_tags)

In [100]:
data['review']

0        one of the other reviewers has mentioned that ...
1        a wonderful little production. the filming tec...
2        i thought this was a wonderful way to spend ti...
3        basically there's a family where a little boy ...
4        petter mattei's "love in the time of money" is...
                               ...                        
49995    i thought this movie did a down right good job...
49996    bad plot, bad dialogue, bad acting, idiotic di...
49997    i am a catholic taught in parochial elementary...
49998    i'm going to have to disagree with the previou...
49999    no one expects the star trek movies to be high...
Name: review, Length: 49582, dtype: object

# Remove Punctuation

Remove "!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~" from text

In [101]:
import string, time
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [102]:
def remove_punc(text):
    for char in excude:
        text = text.replace(char,"")
    return text

# Remove Stop words

#Stopwords such as "I,and,or,etc" remove from text

In [103]:
import re
import nltk

In [104]:
nltk.download("stopwords")

[nltk_data] Downloading package stopwords to C:\Users\Priyanka
[nltk_data]     Sawant\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [105]:
from nltk.corpus import stopwords

In [106]:
from nltk.stem.porter import PorterStemmer

In [107]:
ps = PorterStemmer()

In [108]:
stopwords.words('english')

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [109]:
data.head(3)

review sentiment
0  one of the other reviewers has mentioned that ...  positive
1  a wonderful little production. the filming tec...  positive
2  i thought this was a wonderful way to spend ti...  positive

# Clean the data

In [110]:
# Clean the data
def clean_text(text):
    # Remove HTML tags
    text = re.sub('<.*?>', '', text)
    # Remove non-alphabetic characters and convert to lowercase
    text = re.sub('[^a-zA-Z]', ' ', text).lower()
    # Tokenize the text
    words = nltk.word_tokenize(text)
    # Remove stopwords
    words = [w for w in words if w not in stopwords.words('english')]
    # Stem the words
    stemmer = PorterStemmer()
    words = [stemmer.stem(w) for w in words]
    # Join the words back into a string
    text = ' '.join(words)
    return text

In [111]:
%%time

tqdm.pandas()

data['cleaned_text'] = data['review'].progress_apply(clean_text)

C:\Users\Priyanka Sawant\anaconda3\lib\site-packages\tqdm\std.py:658: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  from pandas import Panel



Wall time: 50min 11s


In [112]:
data.head()

review sentiment  \
0  one of the other reviewers has mentioned that ...  positive   
1  a wonderful little production. the filming tec...  positive   
2  i thought this was a wonderful way to spend ti...  positive   
3  basically there's a family where a little boy ...  negative   
4  petter mattei's "love in the time of money" is...  positive   

                                        cleaned_text  
0  one review mention watch oz episod hook right ...  
1  wonder littl product film techniqu unassum old...  
2  thought wonder way spend time hot summer weeke...  
3  basic famili littl boy jake think zombi closet...  
4  petter mattei love time money visual stun film...

In [113]:
# Save the cleaned DataFrame to an Excel file
data.to_excel(r'test_data_output.xlsx',index=False)

#### Q.2.Given any sentence, perform vector semantics i.e., convert the given dataset into vectors using Bag of Words approach.
#### Q.3. Train a simple classifier (using Scikit-learn e.g.: SVM) to perform  sentiment analysis on the generated dataset.

In [114]:
from sklearn.feature_extraction.text import CountVectorizer

In [115]:
cv = CountVectorizer(max_features=5000)

In [116]:
# Create the Bag of Words model
X = data['cleaned_text']
y = data['sentiment']

In [117]:
X.head(2)

0    one review mention watch oz episod hook right ...
1    wonder littl product film techniqu unassum old...
Name: cleaned_text, dtype: object

In [118]:
y.head(3)

0    positive
1    positive
2    positive
Name: sentiment, dtype: object

In [119]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

y =encoder.fit_transform(y)

In [120]:
y

array([1, 1, 1, ..., 0, 0, 0])

In [121]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=1)

In [122]:
X_train.shape,y_train.shape

((39665,), (39665,))

In [123]:
# Applying Bow

In [124]:
from sklearn.feature_extraction.text import CountVectorizer

In [125]:
cv = CountVectorizer(max_features=3000)

In [126]:
X_train_bow = cv.fit_transform(X_train).toarray()
X_test_bow = cv.transform(X_test).toarray()

In [127]:
X_train_bow.shape

(39665, 3000)

In [128]:
# Train the GaussianNB on BOW 
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

gnb.fit(X_train_bow,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [129]:
y_pred = gnb.predict(X_test_bow)

In [130]:
from sklearn.metrics import accuracy_score,confusion_matrix
accuracy_score(y_test,y_pred)

0.7423616012907129

In [131]:
confusion_matrix(y_test,y_pred)

array([[4324,  709],
       [1846, 3038]], dtype=int64)

In [ ]:
Accuracy using GaussianNB is 74%

In [132]:
# Train the GaussiRandomForest Classifier on BOW 

from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()

rf.fit(X_train_bow,y_train)
y_pred = rf.predict(X_test_bow)
accuracy_score(y_test,y_pred)

0.8401734395482505

Accuracy using RandomForestClassifier is 84%

#### Compare the Model accuracy

We can observe & concluse that RandomForestClassifier is giving better accuracy than GaussianNB.